heartbeat/status/metricsUpdater_test.go

```diff
@@ -0,0 +1,377 @@
+
+		assert.True(t, errors.Is(err, heartbeat.ErrInvalidTimeDuration))
+		assert.True(t, check.IfNil(updater))
+	})
+	t.Run("nil epoch notifier should error", func(t *testing.T) {
+		t.Parallel()
+
+		args := createMockArgsMetricsUpdater()
+		args.EpochNotifier = nil
+		updater, err := NewMetricsUpdater(args)
+
+		assert.Equal(t, heartbeat.ErrNilEpochNotifier, err)
+		assert.True(t, check.IfNil(updater))
+	})
+	t.Run("should work", func(t *testing.T) {
+		t.Parallel()
+
+		args := createMockArgsMetricsUpdater()
+		updater, err := NewMetricsUpdater(args)
+
+		assert.Nil(t, err)
+		assert.False(t, check.IfNil(updater))
+
+		_ = updater.Close()
+	})
+}
+
+func TestMetricsUpdater_Close(t *testing.T) {
+	t.Parallel()
+
+	setUin64Called := atomic.Flag{}
+	args := createMockArgsMetricsUpdater()
+	args.AppStatusHandler = &statusHandler.AppStatusHandlerStub{
+		SetUInt64ValueHandler: func(key string, value uint64) {
+			setUin64Called.SetValue(true)
+		},
+	}
+	updater, _ := NewMetricsUpdater(args)
+
+	time.Sleep(time.Second*3 + time.Millisecond*500)
+	assert.True(t, setUin64Called.IsSet())
+
+	err := updater.Close()
+	assert.Nil(t, err)
+
+	time.Sleep(time.Second)
+	setUin64Called.SetValue(false)
+
+	time.Sleep(time.Second*3 + time.Millisecond*500)
+	assert.False(t, setUin64Called.IsSet())
+}
+
+func TestMetricsUpdater_updateMetrics(t *testing.T) {
+	t.Parallel()
+
+	args := createMockArgsMetricsUpdater()
+	args.EpochNotifier = &epochNotifier.EpochNotifierStub{
+		RegisterNotifyHandlerCalled: func(handler vmcommon.EpochSubscriberHandler) {
+			handler.EpochConfirmed(4, 0)
+		},
+	}
+	t.Run("heartbeat v1 still enabled should not send metrics", func(t *testing.T) {
+		args.HeartbeatV1DisableEpoch = 5
+		args.AppStatusHandler = &statusHandler.AppStatusHandlerStub{
+			SetUInt64ValueHandler: func(key string, value uint64) {
+				assert.Fail(t, "should have not called SetUInt64")
+			},
+		}
+		updater, _ := NewMetricsUpdaterWithoutGoRoutineStart(args)
+
+		updater.updateMetrics()
+	})
+	t.Run("heartbeat v1 is disabled should send connection metrics", func(t *testing.T) {
+		_ = args.PeerAuthenticationCacher.Put([]byte("key1"), "key1", 0)
+		_ = args.PeerAuthenticationCacher.Put([]byte("key2"), "key2", 0)
+		_ = args.PeerAuthenticationCacher.Put([]byte("key3"), "key2", 0)
+
+		args.HeartbeatMonitor = &mock.HeartbeatMonitorStub{
+			GetHeartbeatsCalled: func() []data.PubKeyHeartbeat {
+				return []data.PubKeyHeartbeat{
+					{
+						IsActive: false,
+					},
+					{
+						IsActive: true,
+						PeerType: string(common.EligibleList),
+					},
+					{
+						IsActive: true,
+						PeerType: string(common.WaitingList),
+					},
+					{
+						IsActive: true,
+						PeerType: string(common.JailedList),
+					},
+					{
+						IsActive: true,
+						PeerType: string(common.ObserverList),
+					},
+					{
+						IsActive: true,
+						PeerType: string(common.NewList),
+					},
+				}
+			},
+		}
+		numEpochsToCheck := 4
+		for i := 4; i < numEpochsToCheck+4; i++ {
```

Comment:
for me it is not clear the scope of these for loops.. maybe `"test with epoch %d", i` is confusing, as the `epoch i` is only printed, the test is done 4 times with the same `HeartbeatV1DisableEpoch = 4` and `EpochConfirmed(4, 0)`

Reply:
:facepalm:

Reply:
fixed tests

Reply:
shouldn't it be like this?
```
args.HeartbeatV1DisableEpoch = 4
args.EpochNotifier = &epochNotifier.EpochNotifierStub{
	RegisterNotifyHandlerCalled: func(handler vmcommon.EpochSubscriberHandler) {
		handler.EpochConfirmed(uint32(i), 0)
	},
}
```
with the current implementation, it only updates the metrics once, when `i==4`, thus leading to `updater.shouldSkipUpdateMetrics()` to return true after first iteration

Reply:
will refactor everything, add a new test just for the epoch tests